docs/design/api-versions.md

```diff
@@ -32,8 +40,169 @@ For each API spec, generated `AssignPropertiesTo*()` method  will read the value
+
+``` go
+type GeneratedResource interface {
+    // GetSpec returns the spec of the resource ready for 
+    // serialization to ARM
+    GetSpec(scheme runtime.Scheme) (genruntime.ARMTransformer, error)
+
+    // NewStatus returns a new, empty, status instance ready for
+    // population from ARM
+    NewStatus(scheme runtime.Scheme) (genruntime.FromARMConverter, error)
+
+    // SetStatus updates the current resource with the status provided
+    // The instance passed here should always be the one returned by
+    // NewStatus()
+    SetStatus(status genruntime.FromARMConverter) error
+}
+```
+
+The implementation strategy used for API types and Storage types will differ, as detailed below. Conversion between versions of a resource requires access to a `runtime.Scheme` instance, which is why each method has one as a parameter.
+
+### Implementation for API versions
+
+For all API versions of resources, the generated implementation of this interface will work directly with the `Spec` and `Status` of that API type.
+
+``` go
+// GetSpec returns the Spec of this resource
+func (p *Person) GetSpec(_ runtime.Scheme) (genruntime.ARMTransformer, error) {
+    return &p.Spec, nil
+}
+
+// NewStatus returns a new blank status, ready to be populated
+func (p *Person) NewStatus(_ runtime.Scheme) (genruntime.FromARMConverter, error) {
+    return &Person_Status{}, nil
+}
+
+// SetStatus accepts a status and updates this resource
+// Only a status of the expected type is handled; anything else will error
+func (p *Person) SetStatus(status interface{}) error {
+    s, ok := status.(*Person_Status);
+    if !ok {
+        return errors.Errorf(
+            "expected Status of type %T, but received %T", p.Status, status)
+    } 
+
+    p.Status = *s
+    return nil
+}
+```
+
+### Implementation for storage versions
+
+For the hub Storage version of resources, the generated implementation will transparently handle conversion between versions as dictated by the `OriginalVersion()` function described earlier.
+
+We pivot to the original version by using the `OriginalGVK()` function to obtain a new, empty resource from the schema.
+
+``` go
+// GetSpec returns the spec of the current type using 
+// the original API version from when the CR was created
+func (p *Person) GetSpec(scheme runtime.Scheme) (genruntime.ARMTransformer, error) {
+    gvk := p.OriginalGVK()
+    
+    // Use gvk to create an empty resource of the expected version 
+    originalObj, err := scheme.New(gvk)
+    if err != nil {
+        return errors.Wrapf(
+            err, "unable to create requested version %s of resource", gvk)
+    }
+    resource, ok := originalObj.(genruntime.GeneratedResource)
+    if !ok {
+        return errors.Wrapf(
+            err, "unable to convert %T to genruntime.GeneratedResource", originalObj)
+    }
+
+    // Extract the Spec and populate
+    resultSpec := resource.GetSpec()
```

Comment:
This `GetSpec` also takes `scheme`, but I actually wonder if what you've written is nicer and we want 2 methods on two interfaces?

```go
type KubernetesResource interface {
    GetSpec() interface{}
    // That other stuff it already has
}

type StorageResource interface {
    GetARMSpec(scheme runtime.Scheme) (genruntime.ARMTransformer, error)
}
```

This would convey that: there are two different things we're doing, one of which can fail and one of which cannot. Storage types could implement both interfaces

If you don't want to return `interface{}` on the first one and have to cast, we could also have a `GetSpecAsARMTransformer` for the first interface, although I suspect we'll end up wanting `GetSpec() interface` anyway since right now we're doing that via reflection I'm pretty sure.

Then the storage types could implement `KubernetesResource` _and_ `StorageResource` (or possibly `StorageResource` can embed `KubernetesResource` and we can then change the reconciler to reconcile on that type rather than `KubernetesResource` which is what it's doing now

Reply:
I don't want to have two different interfaces because that pushes responsibilities onto the reconciler to know whether it's working with an API resource or a Storage resource (because it needs to know which interface to use). The goal is to abstract away those details so that the reconciler doesn't need to worry.

Reply:
Yeah, fair enough.

In retrospect I think that this suggestion was really born from the `GetSpec` name making me think that the impl on the API type was really wanting to be `GetSpec() interface{}` when that's not its intent. A name that more clearly exposes the "Get me a thing that can make me an ARM Spec" helps resolve that, and agree that the goal of a common interface for getting that kind of thing between API and storage types is worthwhile.

Reply:
I'm moving the conversion into the reconciler - which both removes the need for a `scheme` parameter, and avoids confusion here because `GetSpec()` just returns the spec on the resource, no conversion required.